pkg/operator/controller/service-dns/controller.go

```diff
@@ -0,0 +1,226 @@
+package service_dns
+
+import (
+	"context"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	"github.com/openshift/cluster-ingress-operator/pkg/resources/dnsrecord"
+
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"k8s.io/client-go/tools/record"
+
+	corev1 "k8s.io/api/core/v1"
+
+	iov1 "github.com/openshift/api/operatoringress/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const controllerName = "service_dns_controller"
+
+var log = logf.Logger.WithName(controllerName)
+
+// NewUnmanaged creates and returns a controller that watches services that have
+// the "dns.openshift.io/domain" annotation and creates dnsrecord objects for
+// them.  This is an unmanaged controller, which means that the manager does not
+// start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		config:   config,
+		client:   mgr.GetClient(),
+		cache:    mgr.GetCache(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	isServiceNeedingDNS := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		svc := o.(*corev1.Service)
+		return len(svc.Labels["gateway.istio.io/managed"]) != 0
+	})
+	gatewayListenerChanged := predicate.Funcs{
```

Comment:
Some other things could change, like ports, protocol, etc https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.Listener

Reply:
Port and protocol shouldn't matter; the DNS record only changes if the host name changes.

Reply:
I added a comment explaining this in https://github.com/openshift/cluster-ingress-operator/compare/9742e1ad86e7114b21eb78fde9be83caccc596b3..d4ac3988e55abd6c35f865e03366a20245ad4256.